<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>An R Toolkit for Evaluating Gene Signatures as Phenotypic Markers • markeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="An R Toolkit for Evaluating Gene Signatures as Phenotypic Markers">
<meta name="description" content="markeR provides a suite of methods for using gene sets (signatures) to quantify and evaluate the extent to which a given gene signature marks a specific phenotype. The package implements various scoring, enrichment and classification approaches, along with tools to compute performance metrics and visualize results, making it a valuable resource for transcriptomics research (bulk RNA-seq).">
<meta property="og:description" content="markeR provides a suite of methods for using gene sets (signatures) to quantify and evaluate the extent to which a given gene signature marks a specific phenotype. The package implements various scoring, enrichment and classification approaches, along with tools to compute performance metrics and visualize results, making it a valuable resource for transcriptomics research (bulk RNA-seq).">
<meta property="og:image" content="https://diseasetranscriptomicslab.github.io/markeR/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">markeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/Tutorial_BenchmarkingMode.html">Benchmarking Mode Tutorial</a></li>
    <li><a class="dropdown-item" href="articles/Tutorial_DiscoveryMode.html">Discovery Mode Tutorial</a></li>
    <li><a class="dropdown-item" href="articles/Tutorial_SignatureSimilarity.html">Signature Similarity Tutorial</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseTranscriptomicsLab/markeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="marker-">markeR 
<a class="anchor" aria-label="anchor" href="#marker-"></a>
</h1>
</div>
<!-- badges: start -->

<blockquote>
<p><strong>To cite markeR please use:</strong></p>
<p>Martins-Silva R, Kaizeler A, Barbosa-Morais N (2025). <em>markeR: an R Toolkit for Evaluating Gene Sets as Phenotypic Markers</em>. Gulbenkian Institute for Molecular Medicine, Faculdade de Medicina, Universidade de Lisboa, Lisbon, Portugal. R package version 0.99.0, <a href="https://github.com/DiseaseTranscriptomicsLab/markeR" class="external-link uri">https://github.com/DiseaseTranscriptomicsLab/markeR</a>.</p>
</blockquote>
<p><strong>markeR</strong> provides a suite of methods for using gene sets (signatures) to quantify and evaluate the extent to which a given gene signature marks a specific phenotype from gene expression data. The package implements various scoring, enrichment and classification approaches, along with tools to compute performance metrics and visualize results.</p>
<p><img src="reference/figures/Workflow.png"></p>
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ul>
<li>
<a href="#installation">Installation</a><br>
</li>
<li>
<a href="#tutorials">Tutorials</a><br>
</li>
<li>
<a href="#common-workflow">Common Workflow</a>
<ul>
<li>
<a href="#id_1-input-requirements">1. Input Requirements</a><br>
</li>
<li>
<a href="#id_2-select-mode-of-analysis">2. Select Mode of Analysis</a><br>
</li>
<li>
<a href="#id_3-choose-a-quantification-approach">3. Choose a Quantification Approach</a>
<ul>
<li>
<a href="#id_31-score-based-approach">3.1 Score-Based Approach</a><br>
</li>
<li>
<a href="#id_32-enrichment-based-approach">3.2 Enrichment-Based Approach</a><br>
</li>
</ul>
</li>
<li>
<a href="#id_4-visualisation-and-evaluation">4. Visualisation and Evaluation</a><br>
</li>
<li>
<a href="#id_5-individual-gene-exploration-optional-but-recommended">5. Individual Gene Exploration (Optional)</a><br>
</li>
<li>
<a href="#id_6-optional-compare-with-reference-gene-sets">6. Compare with Reference Gene Sets (Optional)</a><br>
</li>
</ul>
</li>
<li><a href="#contact">Contact</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The user can install the development version of markeR from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"DiseaseTranscriptomicsLab/markeR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<p>The following tutorials are available:</p>
<ul>
<li><a href="https://diseasetranscriptomicslab.github.io/markeR/articles/benchmarking_tutorial.html">Benchmarking Mode</a></li>
<li><a href="https://diseasetranscriptomicslab.github.io/markeR/articles/discovery_tutorial.html">Discovery Mode</a></li>
<li><a href="https://diseasetranscriptomicslab.github.io/markeR/articles/signaturesimilarity_tutorial.html">Signature Similarity</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="common-workflow">Common Workflow<a class="anchor" aria-label="anchor" href="#common-workflow"></a>
</h2>
<p><code>markeR</code> provides a modular pipeline to quantify transcriptomic signatures and assess their association with phenotypic or clinical variables. The typical workflow includes the following steps:</p>
<div class="section level3">
<h3 id="id_1-input-requirements">1. Input Requirements<a class="anchor" aria-label="anchor" href="#id_1-input-requirements"></a>
</h3>
<p>Depending on the analysis mode, inputs vary slightly.</p>
<ul>
<li>
<p><strong>Gene Set(s)</strong>:<br>
A named list where each element represents one gene set:</p>
<ul>
<li>Use a <strong>character vector</strong> for gene sets where direction of enrichment is not known.</li>
<li>Use a <strong>data frame</strong> with gene names and a directionality column (<code>-1</code> for down-regulated, <code>+1</code> for up-regulated)</li>
</ul>
</li>
</ul>
<p>This structure supports both <strong>Discovery Mode</strong> (single gene set) and <strong>Benchmarking Mode</strong> (multiple gene sets).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example </span></span>
<span><span class="va">gene_sets</span></span>
<span><span class="co">#&gt; $Set1</span></span>
<span><span class="co">#&gt; [1] "GeneA" "GeneB" "GeneC" "GeneD"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Set2</span></span>
<span><span class="co">#&gt;    gene direction</span></span>
<span><span class="co">#&gt; 1 GeneX         1</span></span>
<span><span class="co">#&gt; 2 GeneY        -1</span></span>
<span><span class="co">#&gt; 3 GeneZ         1</span></span></code></pre></div>
<ul>
<li>
<strong>Expression Data Frame</strong>:<br>
A filtered and normalised gene expression data frame (genes × samples). Row names must be gene identifiers, and column names must match the sample IDs in the metadata.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">expr_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Sample1  Sample2  Sample3  Sample4  Sample5</span></span>
<span><span class="co">#&gt; GeneA 3.879049 7.448164 2.864353 5.852928 3.610586</span></span>
<span><span class="co">#&gt; GeneB 4.539645 5.719628 4.564050 4.409857 4.584165</span></span>
<span><span class="co">#&gt; GeneC 8.117417 5.801543 2.947991 6.790251 2.469207</span></span>
<span><span class="co">#&gt; GeneD 5.141017 5.221365 3.542218 6.756267 9.337912</span></span>
<span><span class="co">#&gt; GeneX 5.258575 3.888318 3.749921 6.643162 7.415924</span></span>
<span><span class="co">#&gt; GeneY 8.430130 8.573826 1.626613 6.377281 2.753783</span></span></code></pre></div>
<ul>
<li>
<strong>Sample Metadata</strong>:<br>
A data frame with annotations for each sample, with the sample ID in the first column. The row names must match the column names of the expression matrix.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span></span>
<span><span class="co">#&gt;   SampleID Condition Age</span></span>
<span><span class="co">#&gt; 1  Sample1   Control  49</span></span>
<span><span class="co">#&gt; 2  Sample2 Treatment  44</span></span>
<span><span class="co">#&gt; 3  Sample3   Control  46</span></span>
<span><span class="co">#&gt; 4  Sample4 Treatment  49</span></span>
<span><span class="co">#&gt; 5  Sample5   Control  38</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_2-select-mode-of-analysis">2. Select Mode of Analysis<a class="anchor" aria-label="anchor" href="#id_2-select-mode-of-analysis"></a>
</h3>
<ul>
<li><p><strong>Discovery Mode</strong>: Explore how a single, well-characterised gene set relates to a specific variable of interest. Suitable for hypothesis generation or signature projection.</p></li>
<li><p><strong>Benchmarking Mode</strong>: Evaluate one or more gene sets against multiple metadata variables using a standardised scoring and effect size framework. This mode provides comprehensive visualisations and comparisons across methods.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="id_3-choose-a-quantification-approach">3. Choose a Quantification Approach<a class="anchor" aria-label="anchor" href="#id_3-choose-a-quantification-approach"></a>
</h3>
<p><code>markeR</code> supports two complementary strategies for quantifying the association between gene sets and phenotypes:</p>
<div class="section level4">
<h4 id="id_31-score-based-approach">3.1 Score-Based Approach<a class="anchor" aria-label="anchor" href="#id_31-score-based-approach"></a>
</h4>
<p>This strategy generates a <strong>single numeric score per sample</strong>, reflecting the activity of a gene set. It enables flexible downstream analyses, including comparisons across phenotypic groups.</p>
<p>Three scoring methods are available:</p>
<ul>
<li><p><strong>Log2-median</strong>: Calculates the median log2 expression of the genes in the set. Sensitive to absolute shifts in expression.</p></li>
<li><p><strong>Ranking</strong>: Ranks all genes within each sample and averages the ranks of gene set members. Captures relative ordering rather than magnitude.</p></li>
<li><p><strong>ssGSEA</strong>: Computes a single-sample gene set enrichment score using the ssGSEA algorithm. Reflects the coordinated up- or down-regulation of the set in each sample.</p></li>
</ul>
<p>These methods vary in assumptions and sensitivity. Robust gene sets are expected to perform consistently across all three.</p>
</div>
<div class="section level4">
<h4 id="id_32-enrichment-based-approach">3.2 Enrichment-Based Approach<a class="anchor" aria-label="anchor" href="#id_32-enrichment-based-approach"></a>
</h4>
<p>This approach uses a classical <strong>gene set enrichment analysis (GSEA)</strong> framework to evaluate whether the gene set is significantly overrepresented at the top or bottom of a ranked list of genes (e.g., ranked by fold change or correlation with phenotype).</p>
<ul>
<li>
<strong>GSEA</strong>: Computes a Normalised Enrichment Score (NES) for each contrast or variable of interest, adjusting for gene set size and multiple testing.</li>
</ul>
<p>Use this approach when interested in collective behaviour of gene sets in relation to ranked differential signals.</p>
</div>
</div>
<div class="section level3">
<h3 id="id_4-visualisation-and-evaluation">4. Visualisation and Evaluation<a class="anchor" aria-label="anchor" href="#id_4-visualisation-and-evaluation"></a>
</h3>
<p>In <strong>Benchmarking Mode</strong>, <code>markeR</code> offers a range of visual summaries:</p>
<ul>
<li>Violin or scatter plots showing score distributions by phenotype</li>
<li>Volcano plots and heatmaps based on effect sizes (Cohen’s <em>d</em> or <em>f</em>)</li>
<li>ROC curves and AUC values</li>
<li>Null distribution testing using random gene sets matched for size and directionality</li>
</ul>
<p>In <strong>Discovery Mode</strong>, the output focuses on a single gene set:</p>
<ul>
<li>Score distributions by phenotype</li>
<li>Pairwise contrasts (Cohen’s <em>d</em>) and overall effect sizes (Cohen’s <em>f</em>)</li>
<li>Enrichment score summaries (NES) with adjusted p-values (e.g., lollipop plots)</li>
</ul>
<p>Benchmarking mode offers the most comprehensive set of features and allows users to seamlessly move from discovery to benchmarking mode once a variable of interest has been identified and further testing is required. The main difference from Discovery mode is that Benchmarking is designed to evaluate multiple gene sets simultaneously, whereas Discovery mode focuses on quantifying a single, robust gene set.</p>
</div>
<div class="section level3">
<h3 id="id_5-individual-gene-exploration-optional">5. Individual Gene Exploration (Optional)<a class="anchor" aria-label="anchor" href="#id_5-individual-gene-exploration-optional"></a>
</h3>
<p>To better understand the contribution of individual genes within a gene set and identify whether specific genes drive the overall signal, <code>markeR</code> offers a suite of gene-level exploratory analyses, including:</p>
<ul>
<li>Expression heatmaps of genes across samples and groups</li>
<li>Violin plots showing expression distributions of individual genes</li>
<li>Correlation heatmaps to reveal co-expression patterns among genes in the set</li>
<li>ROC curves and AUC values for individual genes to evaluate their discriminatory power</li>
<li>Effect size calculations (Cohen’s <em>d</em>) per gene to quantify differential expression</li>
<li>Principal Component Analysis (PCA) on gene set genes to assess variance explained and sample clustering</li>
</ul>
</div>
<div class="section level3">
<h3 id="id_6-compare-with-reference-gene-sets-optional">6. Compare with Reference Gene Sets (Optional)<a class="anchor" aria-label="anchor" href="#id_6-compare-with-reference-gene-sets-optional"></a>
</h3>
<p><code>markeR</code> allows comparison of user-defined gene sets to reference sets (e.g., from MSigDB) using:</p>
<ul>
<li><p><strong>Jaccard Index</strong>: Measures gene overlap relative to union size.</p></li>
<li><p><strong>Log Odds Ratio (logOR)</strong>: Computes enrichment using a user-defined gene universe and Fisher’s exact test.</p></li>
</ul>
<p>Filters can be applied based on similarity thresholds (e.g., minimum Jaccard, OR, or p-value).</p>
<hr>
<p>📩 For any questions or concerns, feel free to reach out:</p>
<p><strong>Rita Martins-Silva</strong><br>
Email: <a href="mailto:rita.silva@gimm.pt" class="email">rita.silva@gimm.pt</a></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/DiseaseTranscriptomicsLab/markeR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/DiseaseTranscriptomicsLab/markeR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing markeR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Rita Martins-Silva <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-1067-7993" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Alexandre Kaizeler <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0002-9117-6073" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Nuno Luís Barbosa-Morais <br><small class="roles"> Author, lead, thesis advisor </small> <a href="https://orcid.org/0000-0002-1215-0538" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/DiseaseTranscriptomicsLab/markeR/actions/workflows/r.yaml" class="external-link"><img src="https://github.com/DiseaseTranscriptomicsLab/markeR/actions/workflows/r.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://diseasetranscriptomicslab.github.io/markeR/"><img src="https://img.shields.io/badge/docs-pkgdown-blue.svg" alt="Pkgdown"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rita Martins-Silva, Alexandre Kaizeler, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
