---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# markeR

<!-- badges: start -->
![Development Status](https://img.shields.io/badge/status-development-yellowgreen)
<!-- badges: end -->


**markeR** provides a suite of methods for using gene sets (signatures) to quantify and evaluate the extent to which a given gene signature marks a specific phenotype. The package implements various scoring, enrichment and classification approaches, along with tools to compute performance metrics and visualize results, making it a valuable resource for transcriptomics research (bulk RNA-seq).

## Installation

You can install the development version of markeR from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("DiseaseTranscriptomicsLab/markeR")
```

<!-- ## Methods -->

<!-- The package implements three main approaches for gene signature analysis: -->

<!-- **1. Score-based Methods:**   -->

<!-- These approaches compute a numerical score representing the expression of a gene signature per sample. -->

<!-- - **Log2 Median-Centered Scoring:** Measures signature expression by centering expression values relative to their median. -->
<!-- - **Single-Sample Gene Set Enrichment Analysis (ssGSEA):** Calculates an enrichment score for a given gene set in each sample. -->

<!-- **2. Enrichment-based Methods:**   -->

<!-- These methods evaluate whether a gene signature is significantly enriched in a ranked gene list. -->

<!-- - **Gene Set Enrichment Analysis (GSEA):** A rank-based method for testing enrichment significance in predefined gene sets. -->

<!-- **3. Classification-based Methods:**   -->

<!-- These approaches classify samples into phenotypic groups based on gene signature expression. -->

<!-- - **Random Forest Classification:** Uses decision trees to predict sample labels based on gene signature scores. -->

## Main Functions and Future Modules

The current release of **markeR** includes two primary functions for score-based analysis:

- **CalculateScores:** Calculates gene signature scores for each sample using either the ssGSEA or log2 median-centered method.
- **PlotScores:** Visualizes the calculated scores across conditions using violin plots.

It also includes some functions for visualising individual genes from a gene signature:

- **IndividualGenes_Violins:** creates violin plots of gene expression data with jittered points and optional faceting, allowing for visualization of individual gene expression distributions across sample groups.
- **CorrelationHeatmap:**  computes and visualizes a correlation heatmap for a given set of genes. Optionally, the heatmap can be generated separately for different conditions based on metadata.

Future updates will expand the package with additional pairs of functions to:

- Perform enrichment-based analysis (to calculate and visualize GSEA statistics).
- Conduct classification-based analysis (to train classifiers, e.g., using Random Forests, and to evaluate performance via ROC curves).
- Provide additional gene-level visualization modules to display expression patterns of individual genes within signatures.


## Example

This example demonstrates the calculation of a log2-median-centered score using mock RNA-seq expression data. The dataset is derived from the Marthandan et al. (2016) study (GSE63577) and includes fibroblast samples under replicative senescent and proliferative conditions. (see `?counts_example` and `?metadata_example` for more details).This example showcases how to compute gene expression scores by applying the log2-median-centered transformation.

We are using as an example a very simple senescence gene signature, composed of the genes usually associated with senescence.

```{r example}
library(markeR)
```

```{r}
# Define simple Senescence Signature
SimpleSenescenceSignature <- c("CDKN1A", "CDKN2A", "GLB1","TP53","CCL2")
```

```{r}
data(metadata_example)
data(counts_example)

# Load example data
head(metadata_example)
counts_example[1:5,1:5]
```

### Visualise Individual Genes from Senescence Signature

```{r exampleviolins, fig.width=6, fig.height=3, out.width="70%", warning=FALSE, message=FALSE}

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)


IndividualGenes_Violins(data = counts_example, 
                        metadata = metadata_example, 
                        genes = SimpleSenescenceSignature, 
                        GroupingVariable = "Condition", 
                        plot=T, 
                        ncol=NULL, 
                        nrow=1, 
                        divide=NULL, 
                        invert_divide=FALSE,
                        ColorValues=senescence_triggers_colors, 
                        pointSize=2, 
                        ColorVariable="SenescentType", 
                        title="Senescence Genes", 
                        widthTitle=16,
                        y_limits = NULL,
                        legend_nrow=1, 
                        xlab="Condition",
                        colorlab="") 
```


```{r example_heatmap, fig.width=5, fig.height=3, out.width="70%", warning=FALSE, message=FALSE}
CorrelationHeatmap(data=counts_example, 
                   metadata=metadata_example, 
                   genes=SimpleSenescenceSignature, 
                   separate.by="Condition", 
                   method="spearman", 
                   plot=TRUE, 
                   colorlist=list(low="#3F4193",
                                  mid="#F9F4AE",
                                  high="#B44141"), 
                   limits_colorscale=c(-1,1), 
                   widthTitle=16, 
                   ncol=NULL, 
                   nrow=1, 
                   detailedresults=FALSE, 
                   title="Senescence Genes")
```


### Calculate Senescence Scores

The following example uses the **"logmedian"** method for score calculation.

```{r exampleScore, fig.width=4, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "logmedian", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature))

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)
 
plt <- PlotScores(ResultsList = df_Scores, 
                  ColorVariable = "SenescentType", 
                  GroupingVariable="Condition",  
                  method ="logmedian", 
                  ColorValues = senescence_triggers_colors, 
                  ConnectGroups=TRUE, 
                  ncol = NULL, 
                  nrow = NULL, 
                  widthTitle=20, 
                  y_limits = NULL, 
                  legend_nrow = 1, 
                  pointSize=4)  

ggpubr::annotate_figure(plt, top = grid::textGrob("Marthandan et al. 2016", gp = grid::gpar(cex = 1.3, fontsize = 12)))

```

The following example uses the **"ssGSEA"** method for score calculation.


```{r exampleScore_2, fig.width=4, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "ssGSEA", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature))

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)
 
plt <- PlotScores(ResultsList = df_Scores, 
                  ColorVariable = "SenescentType", 
                  GroupingVariable="Condition",  
                  method ="ssGSEA", 
                  ColorValues = senescence_triggers_colors, 
                  ConnectGroups=TRUE, 
                  ncol = NULL, 
                  nrow = NULL, 
                  widthTitle=20, 
                  y_limits = NULL, 
                  legend_nrow = 1, 
                  pointSize=4)  

ggpubr::annotate_figure(plt, top = grid::textGrob("Marthandan et al. 2016", gp = grid::gpar(cex = 1.3, fontsize = 12)))

```
