---
output: github_document 
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# markeR

<!-- badges: start -->
![Development Status](https://img.shields.io/badge/status-development-yellowgreen)
<!-- badges: end -->


**markeR** provides a suite of methods for using gene sets (signatures) to quantify and evaluate the extent to which a given gene signature marks a specific phenotype. The package implements various scoring, enrichment and classification approaches, along with tools to compute performance metrics and visualize results, making it a valuable resource for transcriptomics research (bulk RNA-seq).

## Table of Contents
- [Installation](#installation)
- [Main Functions and Future Modules](#main-functions-and-future-modules)
- [Example](#example)
  - [Visualise Individual Genes from Senescence Signature](#visualise-individual-genes-from-senescence-signature)
    - [Expression Heatmap](#expression-heatmap)
    - [Expression Violins](#expression-violins)
    - [Correlation Heatmap](#correlation-heatmap)
    - [ROC and AUC](#roc-and-auc)
    - [Cohen's D](#cohens-d)
    - [PCA with only genes of interest](#pca-with-only-genes-of-interest)
  - [Calculate Senescence Scores](#calculate-senescence-scores)
    - [logmedian method](#logmedian-method)
    - [ssGSEA method](#ssgsea-method) 
    - [Ranking method](#ranking-method) 

## Installation

You can install the development version of markeR from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("DiseaseTranscriptomicsLab/markeR")
```

<!-- ## Methods -->

<!-- The package implements three main approaches for gene signature analysis: -->

<!-- **1. Score-based Methods:**   -->

<!-- These approaches compute a numerical score representing the expression of a gene signature per sample. -->

<!-- - **Log2 Median-Centered Scoring:** Measures signature expression by centering expression values relative to their median. -->
<!-- - **Single-Sample Gene Set Enrichment Analysis (ssGSEA):** Calculates an enrichment score for a given gene set in each sample. -->

<!-- **2. Enrichment-based Methods:**   -->

<!-- These methods evaluate whether a gene signature is significantly enriched in a ranked gene list. -->

<!-- - **Gene Set Enrichment Analysis (GSEA):** A rank-based method for testing enrichment significance in predefined gene sets. -->

<!-- **3. Classification-based Methods:**   -->

<!-- These approaches classify samples into phenotypic groups based on gene signature expression. -->

<!-- - **Random Forest Classification:** Uses decision trees to predict sample labels based on gene signature scores. -->

## Main Functions and Future Modules

The current release of **markeR** includes two primary functions for score-based analysis:

- **CalculateScores:** Calculates gene signature scores for each sample using either the ssGSEA, log2 median-centered or ranking method.
- **PlotScores:** Visualizes the calculated scores across conditions using violin plots.

It also includes some functions for visualising individual genes from a gene signature:

- **IndividualGenes_Violins:** creates violin plots of gene expression data with jittered points and optional faceting, allowing for visualization of individual gene expression distributions across sample groups.
- **CorrelationHeatmap:**  computes and visualizes a correlation heatmap for a given set of genes. Optionally, the heatmap can be generated separately for different conditions based on metadata.
- **ExpressionHeatmap:** generates an expression heatmap with customizable sample annotations for a given set of genes.
- **ROCandAUCplot:** computes ROC curves and AUC values for each gene based on gene expression data and sample metadata. It can generate ROC plots, an AUC heatmap, or both arranged side‐by‐side.
- **CohenDHeatmap:** computes Cohen's d for each gene based on gene expression data and sample metadata. The resulting effect sizes are then visualized as a heatmap.
- **plotPCA:** performs PCA on a given dataset and visualizes the results using ggplot2. It allows users to specify genes of interest (to understand if they are sufficient to explain the main variance in the data), customize scaling and centering, and color points based on a metadata variable.

Future updates will expand the package with additional pairs of functions to:

- Perform enrichment-based analysis (to calculate and visualize GSEA statistics).
- Conduct classification-based analysis (to train classifiers, e.g., using Random Forests, and to evaluate performance via ROC curves).
- Provide additional gene-level visualization modules to display expression patterns of individual genes within signatures.


## Example

This example demonstrates the calculation of a log2-median-centered score using mock RNA-seq expression data. The dataset is derived from the Marthandan et al. (2016) study (GSE63577) and includes fibroblast samples under replicative senescent and proliferative conditions. (see `?counts_example` and `?metadata_example` for more details).This example showcases how to compute gene expression scores by applying the log2-median-centered transformation.

We are using as an example a very simple senescence gene signature, composed of the genes usually associated with senescence.

```{r example}
library(markeR)
```

```{r}
# Define simple Senescence Signature
SimpleSenescenceSignature <- c("CDKN1A", "CDKN2A", "GLB1","TP53","CCL2", "LMNB1", "MKI67" )
```

```{r}
data(metadata_example)
data(counts_example)

# Load example data
head(metadata_example)
counts_example[1:5,1:5]
```

### Visualise Individual Genes from Senescence Signature

#### Expression Heatmap

```{r example_exprheatmap, fig.width=8, fig.height=3, out.width="70%", warning=FALSE, message=FALSE}

annotation_colors <- list( 
  Condition = c(
    "Senescent"     = "#65AC7C",  # Example color: greenish
    "Proliferative" = "#5F90D4"  # Example color: blueish 
  )
)

ExpressionHeatmap(data=counts_example, 
                  metadata = metadata_example, 
                  genes=SimpleSenescenceSignature,  
                  annotate.by = c("Condition"),
                  annotation_colors = annotation_colors,
                  colorlist = list(low = "#3F4193", mid = "#F9F4AE", high = "#B44141"),
                  cluster_rows = TRUE, 
                  cluster_columns = FALSE,
                  title = "Senescence Genes", 
                  titlesize = 20,
                  legend_position = "right",
                  scale_position="right")
```
#### Expression Violins

```{r exampleviolins, fig.width=10, fig.height=3, out.width="100%", warning=FALSE, message=FALSE}

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)


IndividualGenes_Violins(data = counts_example, 
                        metadata = metadata_example, 
                        genes = SimpleSenescenceSignature, 
                        GroupingVariable = "Condition", 
                        plot=T, 
                        ncol=NULL, 
                        nrow=1, 
                        divide=NULL, 
                        invert_divide=FALSE,
                        ColorValues=senescence_triggers_colors, 
                        pointSize=2, 
                        ColorVariable="SenescentType", 
                        title="Senescence Genes", 
                        widthTitle=16,
                        y_limits = NULL,
                        legend_nrow=1, 
                        xlab="Condition",
                        colorlab="") 
```

#### Correlation Heatmap

```{r example_heatmap, fig.width=8, fig.height=4, out.width="70%", warning=FALSE, message=FALSE}

CorrelationHeatmap(data=counts_example, 
                   metadata = metadata_example, 
                   genes=SimpleSenescenceSignature, 
                   separate.by = "Condition", 
                   method = "spearman",  
                   colorlist = list(low = "#3F4193", mid = "#F9F4AE", high = "#B44141"),
                   limits_colorscale = c(-1,0,1), 
                   widthTitle = 16, 
                   title = "Senescence Genes", 
                   cluster_rows = TRUE, 
                   cluster_columns = TRUE,  
                   detailedresults = FALSE, 
                   legend_position="right",
                   titlesize=20)
```

#### ROC and AUC 

```{r rocAUCexample, fig.width=12, fig.height=4, out.width="100%", warning=FALSE, message=FALSE}

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)
 
ROCandAUCplot(counts_example, 
              metadata_example, 
              condition_var = "Condition", 
              class = "Senescent", 
              group_var=NULL,
              genes=SimpleSenescenceSignature, 
              plot_type = "all",
              heatmap_params = list(col = list( "#F9F4AE" ,"#B44141"),
                                    limits = c(0.5,1),
                                    cluster_rows=T),
              roc_params = list(nrow=3,
                                ncol=3,
                                colors=senescence_triggers_colors),
              commomplot_params = list(widths=c(0.5,0.3)))
```

#### Cohen's D

```{r cohendexample, fig.width=6, fig.height=4, out.width="60%", warning=FALSE, message=FALSE}
CohenDHeatmap(counts_example, 
              metadata_example, 
              genes=SimpleSenescenceSignature,
              condition_var = "Condition", 
              class = "Senescent", 
              group_var = NULL,  
              heatmap_params = list(col = list( "#F9F4AE" ,"#B44141"),
                                    limits = NULL,
                                    cluster_rows=T))
```

#### PCA with only genes of interest


```{r pca, fig.width=7, fig.height=3, out.width="90%", warning=FALSE, message=FALSE}

annotation_colors <- c(  
    "Senescent"     = "#65AC7C",  # Example color: greenish
    "Proliferative" = "#5F90D4"  # Example color: blueish 
)

 
plotPCA(data = counts_example, 
        metadata = metadata_example, 
        genes=SimpleSenescenceSignature, 
        scale=FALSE, 
        center=TRUE, 
        PCs=list(c(1,2), c(2,3), c(3,4)), 
        ColorVariable="Condition",
        ColorValues=annotation_colors,
        pointSize=5,
        legend_nrow=1, 
        ncol=3, 
        nrow=NULL)
```



### Calculate Senescence Scores

#### logmedian method

The following example uses the **"logmedian"** method for score calculation.

```{r exampleScore, fig.width=3, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "logmedian", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature))

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)

cond_cohend <- list(A=c("Senescent"),  
                    B=c("Proliferative"))

PlotScores(ResultsList = df_Scores, 
           ColorVariable = "SenescentType", 
           GroupingVariable="Condition",  
           method ="logmedian", 
           ColorValues = senescence_triggers_colors, 
           ConnectGroups=TRUE, 
           ncol = NULL, 
           nrow = NULL, 
           widthTitle=24, 
           y_limits = NULL, 
           legend_nrow = 1, 
           pointSize=4,
           cond_cohend=cond_cohend,
           title="Marthandan et al. 2016",
           labsize=7, 
           titlesize = 10)  

```

Given that some of the genes are expected to be upregulated, while others to be downregulated in senescence, we can also consider a *bidirectional signature*.

```{r exampleScore_bidirectional, fig.width=3, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
SimpleSenescenceSignature_bidirectional <- data.frame(gene=c("CDKN1A", "CDKN2A", "GLB1","TP53","CCL2", "LMNB1", "MKI67" ),
                                                      enrichment=c(1,1,1,1,1,-1,-1))

df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "logmedian", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature_bidirectional))

PlotScores(ResultsList = df_Scores, 
           ColorVariable = "SenescentType", 
           GroupingVariable="Condition",  
           method ="logmedian", 
           ColorValues = senescence_triggers_colors, 
           ConnectGroups=TRUE, 
           ncol = NULL, 
           nrow = NULL, 
           widthTitle=24, 
           y_limits = NULL, 
           legend_nrow = 1, 
           pointSize=4,
           cond_cohend=cond_cohend,
           title="Marthandan et al. 2016",
           labsize=7, 
           titlesize = 10)  

```

#### ssGSEA method

The following example uses the **"ssGSEA"** method for score calculation, both for unidirectional and bidirectional signatures.


```{r exampleScoresGSEA_uni, fig.width=3, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
 
df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "ssGSEA", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature))

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)

cond_cohend <- list(A=c("Senescent"),  
                    B=c("Proliferative"))

PlotScores(ResultsList = df_Scores, 
           ColorVariable = "SenescentType", 
           GroupingVariable="Condition",  
           method ="ssGSEA", 
           ColorValues = senescence_triggers_colors, 
           ConnectGroups=TRUE, 
           ncol = NULL, 
           nrow = NULL, 
           widthTitle=24, 
           y_limits = NULL, 
           legend_nrow = 1, 
           pointSize=4,
           cond_cohend=cond_cohend,
           title="Marthandan et al. 2016",
           labsize=7, 
           titlesize = 10)  



```


```{r examplessGSEA_bi, fig.width=3, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
 
df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "ssGSEA", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature_bidirectional))

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)

cond_cohend <- list(A=c("Senescent"),  
                    B=c("Proliferative"))

PlotScores(ResultsList = df_Scores, 
           ColorVariable = "SenescentType", 
           GroupingVariable="Condition",  
           method ="ssGSEA", 
           ColorValues = senescence_triggers_colors, 
           ConnectGroups=TRUE, 
           ncol = NULL, 
           nrow = NULL, 
           widthTitle=24, 
           y_limits = NULL, 
           legend_nrow = 1, 
           pointSize=4,
           cond_cohend=cond_cohend,
           title="Marthandan et al. 2016",
           labsize=7, 
           titlesize = 10)  



```

#### Ranking method

The following example uses the **"ranking"** method for score calculation, both for unidirectional and bidirectional signatures.



```{r ranking_unidirect, fig.width=3, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
 
df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "ranking", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature))

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)

cond_cohend <- list(A=c("Senescent"),  
                    B=c("Proliferative"))

PlotScores(ResultsList = df_Scores, 
           ColorVariable = "SenescentType", 
           GroupingVariable="Condition",  
           method ="ranking", 
           ColorValues = senescence_triggers_colors, 
           ConnectGroups=TRUE, 
           ncol = NULL, 
           nrow = NULL, 
           widthTitle=24, 
           y_limits = NULL, 
           legend_nrow = 1, 
           pointSize=4,
           cond_cohend=cond_cohend,
           title="Marthandan et al. 2016",
           labsize=7, 
           titlesize = 10)  



```


```{r ranking_bidirect, fig.width=3, fig.height=4, out.width="40%", warning=FALSE, message=FALSE}
 
df_Scores <- CalculateScores(data = counts_example, 
                             metadata = metadata_example, 
                             method = "ranking", 
                             gene_sets = list(Senescence=SimpleSenescenceSignature_bidirectional))

senescence_triggers_colors <- c(
  "none" = "#E57373",  # Soft red   
  "Telomere shortening" = "#4FC3F7"  # Vivid sky blue  
)

cond_cohend <- list(A=c("Senescent"),  
                    B=c("Proliferative"))

PlotScores(ResultsList = df_Scores, 
           ColorVariable = "SenescentType", 
           GroupingVariable="Condition",  
           method ="ranking", 
           ColorValues = senescence_triggers_colors, 
           ConnectGroups=TRUE, 
           ncol = NULL, 
           nrow = NULL, 
           widthTitle=24, 
           y_limits = NULL, 
           legend_nrow = 1, 
           pointSize=4,
           cond_cohend=cond_cohend,
           title="Marthandan et al. 2016",
           labsize=7, 
           titlesize = 10)  



```
