<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate Gene Signature Scores using Score-Based Approaches — CalculateScores • markeR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate Gene Signature Scores using Score-Based Approaches — CalculateScores"><meta name="description" content="This function calculates a gene signature score for each sample based on one or more predefined gene sets
(signatures)."><meta property="og:description" content="This function calculates a gene signature score for each sample based on one or more predefined gene sets
(signatures)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">markeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseTranscriptomicsLab/markeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate Gene Signature Scores using Score-Based Approaches</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseTranscriptomicsLab/markeR/blob/main/R/CalculateScores.R" class="external-link"><code>R/CalculateScores.R</code></a></small>
      <div class="d-none name"><code>CalculateScores.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates a gene signature score for each sample based on one or more predefined gene sets
(signatures).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CalculateScores</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  <span class="va">gene_sets</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ssGSEA"</span>, <span class="st">"logmedian"</span>, <span class="st">"ranking"</span>, <span class="st">"all"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame of normalized (non-transformed) counts where each row is a gene and each column is a sample. The row names should contain gene names,
and the column names should contain sample identifiers. <strong>(Required)</strong></p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>A data frame describing the attributes of each sample. Each row corresponds to a sample and each column to an attribute.
The first column of <code>metadata</code> should be the sample identifiers (i.e., the column names of <code>data</code>).
Defaults to <code>NULL</code> if no metadata is provided.</p></dd>


<dt id="arg-gene-sets">gene_sets<a class="anchor" aria-label="anchor" href="#arg-gene-sets"></a></dt>
<dd><p>Gene set input. <strong>(Required)</strong></p>
<p>If using <strong>unidirectional</strong> gene sets, provide a named list where each element is a vector of gene names representing a gene signature.
The names of the list elements should correspond to the labels for each signature.</p>
<p>If using <strong>bidirectional</strong> gene sets, provide a named list where each element is a data frame. The names of the list elements should
correspond to the labels for each signature, and each data frame should contain the following structure:</p><ul><li><p>The <strong>first column</strong> should contain gene names.</p></li>
<li><p>The <strong>second column</strong> should indicate the expected direction of enrichment (1 for upregulated genes, -1 for downregulated genes).</p></li>
</ul></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string indicating the scoring method to use. Options are <code>"ssGSEA"</code>,
<code>"logmedian"</code>, <code>"ranking"</code>, or <code>"all"</code> (to compute scores using all methods). Defaults to <code>"logmedian"</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If a single method is chosen, a data frame containing the calculated scores for each gene signature, including metadata if provided.
If <code>method = "all"</code>, a list is returned where each element corresponds to a scoring method and contains the respective data frame of scores.</p>
<dl><dt>sample</dt>
<dd><p>The sample identifier (matching the column names of the input data).</p></dd>

<dt>score</dt>
<dd><p>The calculated gene signature score for the corresponding sample.</p></dd>

<dt>(metadata)</dt>
<dd><p>Any additional columns from the <code>metadata</code> data frame provided by the user, if available.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<dl>
This function calculates a gene signature score for each sample based on one or more predefined gene sets
(signatures). Four methods are available:

<dt><code>ssGSEA</code></dt>
<dd><p>Uses the single-sample Gene Set Enrichment Analysis (ssGSEA) method to compute an enrichment score
for each signature in each sample. This method uses an adaptation from the the <code>gsva()</code> function from the <code>GSVA</code> package
to compute an enrichment score, representing the absolute enrichment of each gene set in each sample.</p></dd>

<dt><code>logmedian</code></dt>
<dd><p>Computes, for each sample, the score as the sum of the normalized (log2-median-centered)
expression values of the signature genes divided by the number of genes in the signature.</p></dd>

<dt><code>ranking</code></dt>
<dd><p>Computes gene signature scores for each sample by ranking the expression of signature genes
in the dataset and normalizing the score based on the total number of genes.</p></dd>

<dt><code>all</code></dt>
<dd><p>Computes gene signature scores using all three methods (<code>ssGSEA</code>, <code>logmedian</code>, and <code>ranking</code>).
The function returns a list containing the results of each method.</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span>  <span class="co"># Assume 'gene_data' is your gene expression data frame and 'sample_metadata'</span></span></span>
<span class="r-in"><span>  <span class="co"># is your metadata. Define a list of gene signatures as follows:</span></span></span>
<span class="r-in"><span>  <span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"Signature_A"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene5"</span>, <span class="st">"Gene10"</span>, <span class="st">"Gene20"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="st">"Signature_B"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene2"</span>, <span class="st">"Gene6"</span>, <span class="st">"Gene15"</span>, <span class="st">"Gene30"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># Using the ssGSEA method:</span></span></span>
<span class="r-in"><span>  <span class="va">scores_ssgsea</span> <span class="op">&lt;-</span> <span class="fu">calculate_signature_score</span><span class="op">(</span>data <span class="op">=</span> <span class="va">gene_data</span>,</span></span>
<span class="r-in"><span>                                             metadata <span class="op">=</span> <span class="va">sample_metadata</span>,</span></span>
<span class="r-in"><span>                                             gene_sets <span class="op">=</span> <span class="va">gene_sets</span>,</span></span>
<span class="r-in"><span>                                             method <span class="op">=</span> <span class="st">"ssGSEA"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># Using the logmedian method (default):</span></span></span>
<span class="r-in"><span>  <span class="va">scores_logmedian</span> <span class="op">&lt;-</span> <span class="fu">calculate_signature_score</span><span class="op">(</span>data <span class="op">=</span> <span class="va">gene_data</span>,</span></span>
<span class="r-in"><span>                                                gene_sets <span class="op">=</span> <span class="va">gene_sets</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># Using all methods:</span></span></span>
<span class="r-in"><span>  <span class="va">scores_all</span> <span class="op">&lt;-</span> <span class="fu">calculate_signature_score</span><span class="op">(</span>data <span class="op">=</span> <span class="va">gene_data</span>,</span></span>
<span class="r-in"><span>                                          metadata <span class="op">=</span> <span class="va">sample_metadata</span>,</span></span>
<span class="r-in"><span>                                          gene_sets <span class="op">=</span> <span class="va">gene_sets</span>,</span></span>
<span class="r-in"><span>                                          method <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rita Martins-Silva, Alexandre Kaizeler, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

