<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Cohen's d Heatmap Function — CohenD_IndividualGenes • markeR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cohen's d Heatmap Function — CohenD_IndividualGenes"><meta name="description" content="This function computes Cohen's d for each gene based on gene expression data and sample metadata.
For each gene, it compares the expression values between samples where condition_var equals
class (the positive class) versus the remaining samples. The resulting effect sizes are then
visualized as a heatmap."><meta property="og:description" content="This function computes Cohen's d for each gene based on gene expression data and sample metadata.
For each gene, it compares the expression values between samples where condition_var equals
class (the positive class) versus the remaining samples. The resulting effect sizes are then
visualized as a heatmap."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">markeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseTranscriptomicsLab/markeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Cohen's d Heatmap Function</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseTranscriptomicsLab/markeR/blob/main/R/CohenD_IndividualGenes.R" class="external-link"><code>R/CohenD_IndividualGenes.R</code></a></small>
      <div class="d-none name"><code>CohenD_IndividualGenes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function computes Cohen's d for each gene based on gene expression data and sample metadata.
For each gene, it compares the expression values between samples where <code>condition_var</code> equals
<code>class</code> (the positive class) versus the remaining samples. The resulting effect sizes are then
visualized as a heatmap.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CohenD_IndividualGenes</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">condition_var</span>,</span>
<span>  <span class="va">class</span>,</span>
<span>  group_var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  titlesize <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame or matrix containing gene expression data, with genes as rows and samples as columns.</p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>A data frame containing sample metadata. The first column should contain sample identifiers that match the column names of <code>data</code>.</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>A character vector specifying which genes to include. If <code>NULL</code> (default), all genes in <code>data</code> are used.
A warning is issued if more than 30 genes are selected.</p></dd>


<dt id="arg-condition-var">condition_var<a class="anchor" aria-label="anchor" href="#arg-condition-var"></a></dt>
<dd><p>A character string specifying the column name in <code>metadata</code> representing the condition of interest.
(Mandatory; no default.)</p></dd>


<dt id="arg-class">class<a class="anchor" aria-label="anchor" href="#arg-class"></a></dt>
<dd><p>A character string or vector specifying the positive class label for the condition.
(Mandatory; no default.)</p></dd>


<dt id="arg-group-var">group_var<a class="anchor" aria-label="anchor" href="#arg-group-var"></a></dt>
<dd><p>An optional character string specifying the column name in <code>metadata</code> used for grouping samples.
If not provided (<code>NULL</code>), all samples are treated as a single group.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>An optional character string specifying a custom title for the heatmap.
If not provided, a default title is generated.</p></dd>


<dt id="arg-titlesize">titlesize<a class="anchor" aria-label="anchor" href="#arg-titlesize"></a></dt>
<dd><p>A numeric value specifying the size of the title. Default is <code>14</code>.</p></dd>


<dt id="arg-params">params<a class="anchor" aria-label="anchor" href="#arg-params"></a></dt>
<dd><p>A list of additional parameters for customizing the heatmap. Possible elements include:</p><dl><dt><code>cluster_rows</code></dt>
<dd><p>Logical; if <code>TRUE</code> (default), rows are clustered.</p></dd>

<dt><code>cluster_columns</code></dt>
<dd><p>Logical; if <code>TRUE</code> (default), columns are clustered.</p></dd>

<dt><code>colors</code></dt>
<dd><p>A vector of length 2 of colors to be used for the minimum and maximum values of the color scale.
Defaults to <code>c("#FFFFFF", "#21975C")</code>, but note that the default mapping for Cohen's d is set to a divergent scale.</p></dd>

<dt><code>limits</code></dt>
<dd><p>A numeric vector of length 2 specifying the minimum and maximum values for the color scale.
If not provided, defaults to <code>c(-2, 2)</code>.</p></dd>

<dt><code>name</code></dt>
<dd><p>A character string for the legend title of the color scale. Default is <code>"Cohen's d"</code>.</p></dd>

<dt><code>row_names_gp</code></dt>
<dd><p>Optional graphical parameters for row names (passed to <span class="pkg">ComplexHeatmap</span>).</p></dd>

<dt><code>column_names_gp</code></dt>
<dd><p>Optional graphical parameters for column names (passed to <span class="pkg">ComplexHeatmap</span>).</p></dd>


</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Invisibly returns a list containing:</p><dl><dt><code>plot</code></dt>
<dd><p>The <span class="pkg">ComplexHeatmap</span> object of the Cohen's d heatmap.</p></dd>

<dt><code>data</code></dt>
<dd><p>A data frame with the calculated Cohen's d values for each gene and group.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function computes Cohen's d for each gene by comparing the expression values between samples with
<code>condition_var == class</code> and those that do not. The effect sizes are then visualized as a heatmap using
<span class="pkg">ComplexHeatmap</span>. When <code>group_var</code> is not provided, all samples are treated as a single group.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span>  <span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">CohenDHeatmap</span><span class="op">(</span>data <span class="op">=</span> <span class="va">myData</span>,</span></span>
<span class="r-in"><span>                          metadata <span class="op">=</span> <span class="va">myMetadata</span>,</span></span>
<span class="r-in"><span>                          genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          condition_var <span class="op">=</span> <span class="st">"Genotype"</span>,</span></span>
<span class="r-in"><span>                          class <span class="op">=</span> <span class="st">"Mutant"</span>,</span></span>
<span class="r-in"><span>                          group_var <span class="op">=</span> <span class="st">"CellType"</span>,</span></span>
<span class="r-in"><span>                          title <span class="op">=</span> <span class="st">"Cohen's d Heatmap"</span>,</span></span>
<span class="r-in"><span>                          params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rita Martins-Silva, Alexandre Kaizeler, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

