<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • markeR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">markeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="AUC_Scores.html">AUC_Scores()</a></code>

        </dt>
        <dd>Generate Heatmaps for AUC Scores using ggplot2</dd>
      </dl><dl><dt>

          <code><a href="CalculateScores.html">CalculateScores()</a></code>

        </dt>
        <dd>Calculate Gene Signature Scores using Score-Based Approaches</dd>
      </dl><dl><dt>

          <code><a href="CalculateScores_Ranking.html">CalculateScores_Ranking()</a></code>

        </dt>
        <dd>Calculate Gene Signature Scores using Ranking Approach</dd>
      </dl><dl><dt>

          <code><a href="CalculateScores_logmedian.html">CalculateScores_logmedian()</a></code>

        </dt>
        <dd>Calculate Gene Signature Scores using Log-Median Approach</dd>
      </dl><dl><dt>

          <code><a href="CalculateScores_ssGSEA_bidirectional.html">CalculateScores_ssGSEA_bidirectional()</a></code>

        </dt>
        <dd>Calculate ssGSEA Scores for Bidirectional Gene Signatures</dd>
      </dl><dl><dt>

          <code><a href="CalculateScores_ssGSEA_unidirectional.html">CalculateScores_ssGSEA_unidirectional()</a></code>

        </dt>
        <dd>Calculate ssGSEA Scores for Unidirectional Gene Signatures</dd>
      </dl><dl><dt>

          <code><a href="CohenD_IndividualGenes.html">CohenD_IndividualGenes()</a></code>

        </dt>
        <dd>Cohen's d Heatmap Function</dd>
      </dl><dl><dt>

          <code><a href="CohenD_allConditions.html">CohenD_allConditions()</a></code>

        </dt>
        <dd>Compute Cohen\'s d for All Gene Signatures Across Conditions</dd>
      </dl><dl><dt>

          <code><a href="CohenF_allConditions.html">CohenF_allConditions()</a></code>

        </dt>
        <dd>Compute Cohen's f for All Gene Signatures Across a Categorical Variable</dd>
      </dl><dl><dt>

          <code><a href="CorrelationHeatmap.html">CorrelationHeatmap()</a></code>

        </dt>
        <dd>CorrelationHeatmap: Generate correlation heatmaps with optional grouping</dd>
      </dl><dl><dt>

          <code><a href="ExpressionHeatmap.html">ExpressionHeatmap()</a></code>

        </dt>
        <dd>ExpressionHeatmap: Generate an expression heatmap with customizable sample annotations and separate legend positions</dd>
      </dl><dl><dt>

          <code><a href="FDR_Simulation.html">FDR_Simulation()</a></code>

        </dt>
        <dd>FDR Simulation Plot</dd>
      </dl><dl><dt>

          <code><a href="GSEA_VariableAssociation.html">GSEA_VariableAssociation()</a></code>

        </dt>
        <dd>GSEA Variable Association</dd>
      </dl><dl><dt>

          <code><a href="Heatmap_Cohen.html">Heatmap_Cohen()</a></code>

        </dt>
        <dd>Generate Heatmaps for Cohen\'s d Effect Sizes using ggplot2</dd>
      </dl><dl><dt>

          <code><a href="IndividualGenes_Violins.html">IndividualGenes_Violins()</a></code>

        </dt>
        <dd>Generate Violin Plots for Individual Genes</dd>
      </dl><dl><dt>

          <code><a href="PlotScores.html">PlotScores()</a></code>

        </dt>
        <dd>Plot gene signature scores using various methods.</dd>
      </dl><dl><dt>

          <code><a href="PlotScores_Categorical.html">PlotScores_Categorical()</a></code>

        </dt>
        <dd>Plot Gene Set Scores by Group or Continuous Variable</dd>
      </dl><dl><dt>

          <code><a href="PlotScores_Numeric.html">PlotScores_Numeric()</a></code>

        </dt>
        <dd>Plot Gene Signature Scores with Continuous Variables</dd>
      </dl><dl><dt>

          <code><a href="ROCAUC_Scores_Calculate.html">ROCAUC_Scores_Calculate()</a></code>

        </dt>
        <dd>Compute ROC Curves and AUC Values for Gene Signature Scores</dd>
      </dl><dl><dt>

          <code><a href="ROC_Scores.html">ROC_Scores()</a></code>

        </dt>
        <dd>Plot ROC Curves for Gene Signature Scores</dd>
      </dl><dl><dt>

          <code><a href="ROCandAUCplot.html">ROCandAUCplot()</a></code>

        </dt>
        <dd>ROC and AUC Plot Function</dd>
      </dl><dl><dt>

          <code><a href="Score_VariableAssociation.html">Score_VariableAssociation()</a></code>

        </dt>
        <dd>Score Variable Association</dd>
      </dl><dl><dt>

          <code><a href="VariableAssociation.html">VariableAssociation()</a></code>

        </dt>
        <dd>Visualize Statistical Associations Between Variables and a Target Score</dd>
      </dl><dl><dt>

          <code><a href="VisualiseIndividualGenes.html">VisualiseIndividualGenes()</a></code>

        </dt>
        <dd>Visualize Individual Genes' Expression Using Various Methods</dd>
      </dl><dl><dt>

          <code><a href="Volcano_Cohen.html">Volcano_Cohen()</a></code>

        </dt>
        <dd>Volcano Plot of Cohen's d Effect Sizes and Adjusted p-values</dd>
      </dl><dl><dt>

          <code><a href="calculateDE.html">calculateDE()</a></code>

        </dt>
        <dd>Calculate Differential Gene Expression Statistics using limma</dd>
      </dl><dl><dt>

          <code><a href="colRanking.html">colRanking()</a></code>

        </dt>
        <dd>Compute Independent Column-wise Ranks of Matrix Elements</dd>
      </dl><dl><dt>

          <code><a href="compute_cohens_f_pval.html">compute_cohens_f_pval()</a></code>

        </dt>
        <dd>Compute Cohen's f and p-value for a given model and predictor</dd>
      </dl><dl><dt>

          <code><a href="compute_stat_tests.html">compute_stat_tests()</a></code>

        </dt>
        <dd>Compute Statistical Tests for Variable Associations with a Target Variable</dd>
      </dl><dl><dt>

          <code><a href="counts_example.html">counts_example</a></code>

        </dt>
        <dd>Gene Expression Counts for Marthandan et al. (2016) RNA-Seq Data</dd>
      </dl><dl><dt>

          <code><a href="create_contrast_column.html">create_contrast_column()</a></code>

        </dt>
        <dd>Create Contrast Column in Metadata</dd>
      </dl><dl><dt>

          <code><a href="generate_all_contrasts.html">generate_all_contrasts()</a></code>

        </dt>
        <dd>Generate All Possible Unique Contrasts Between Groups</dd>
      </dl><dl><dt>

          <code><a href="getRanking.html">getRanking()</a></code>

        </dt>
        <dd>Get Gene Expression Ranking</dd>
      </dl><dl><dt>

          <code><a href="identify_variable_type.html">identify_variable_type()</a></code>

        </dt>
        <dd>Identify Variable Types</dd>
      </dl><dl><dt>

          <code><a href="markeR.html">markeR</a></code> <code><a href="markeR.html">markeR-package</a></code>

        </dt>
        <dd>markeR: An R Toolkit for Evaluating Gene Signatures as Phenotypic Markers</dd>
      </dl><dl><dt>

          <code><a href="metadata_example.html">metadata_example</a></code>

        </dt>
        <dd>Metadata for Marthandan et al. (2016) RNA-Seq Study</dd>
      </dl><dl><dt>

          <code><a href="plotCombinedGSEA.html">plotCombinedGSEA()</a></code>

        </dt>
        <dd>Plot Combined GSEA Results</dd>
      </dl><dl><dt>

          <code><a href="plotGSEAenrichment.html">plotGSEAenrichment()</a></code>

        </dt>
        <dd>Plot GSEA Enrichment Results</dd>
      </dl><dl><dt>

          <code><a href="plotNESlollipop.html">plotNESlollipop()</a></code>

        </dt>
        <dd>Create a Lollipop Plot for GSEA Results</dd>
      </dl><dl><dt>

          <code><a href="plotPCA.html">plotPCA()</a></code>

        </dt>
        <dd>Principal Component Analysis (PCA) Plot</dd>
      </dl><dl><dt>

          <code><a href="plotVolcano.html">plotVolcano()</a></code>

        </dt>
        <dd>Volcano Plots from Differential Expression Results</dd>
      </dl><dl><dt>

          <code><a href="remove_division.html">remove_division()</a></code>

        </dt>
        <dd>Remove Division Notation in Contrast Labels</dd>
      </dl><dl><dt>

          <code><a href="runGSEA.html">runGSEA()</a></code>

        </dt>
        <dd>Run Gene Set Enrichment Analysis (GSEA) for Multiple Contrasts</dd>
      </dl><dl><dt>

          <code><a href="ssGSEA_alternative.html">ssGSEA_alternative()</a></code>

        </dt>
        <dd>Alternative Implementation of Single-Sample Gene Set Enrichment Analysis (ssGSEA)</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rita Martins-Silva, Alexandre Kaizeler, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

