<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot Gene Set Scores by Group or Continuous Variable — PlotScores_Categorical • markeR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot Gene Set Scores by Group or Continuous Variable — PlotScores_Categorical"><meta name="description" content="This function computes and visualizes gene set enrichment scores using various methods, optionally comparing across groups or numeric variables.
It supports categorical and numeric comparisons, statistical testing, Cohen's d effect sizes, and visualizations such as heatmaps and volcano plots."><meta property="og:description" content="This function computes and visualizes gene set enrichment scores using various methods, optionally comparing across groups or numeric variables.
It supports categorical and numeric comparisons, statistical testing, Cohen's d effect sizes, and visualizations such as heatmaps and volcano plots."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">markeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseTranscriptomicsLab/markeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot Gene Set Scores by Group or Continuous Variable</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseTranscriptomicsLab/markeR/blob/0.9.0/R/PlotScores.R" class="external-link"><code>R/PlotScores.R</code></a></small>
      <div class="d-none name"><code>PlotScores_Categorical.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function computes and visualizes gene set enrichment scores using various methods, optionally comparing across groups or numeric variables.
It supports categorical and numeric comparisons, statistical testing, Cohen's d effect sizes, and visualizations such as heatmaps and volcano plots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">PlotScores_Categorical</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  <span class="va">gene_sets</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ssGSEA"</span>, <span class="st">"logmedian"</span>, <span class="st">"ranking"</span><span class="op">)</span>,</span>
<span>  ColorVariable <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  GroupingVariable <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ColorValues <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ConnectGroups <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  widthTitle <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  titlesize <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  limits <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pointSize <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  xlab <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  labsize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  compute_cohen <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cond_cohend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pvalcalc <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"simple"</span>, <span class="st">"medium"</span>, <span class="st">"extensive"</span><span class="op">)</span>,</span>
<span>  widthlegend <span class="op">=</span> <span class="fl">22</span>,</span>
<span>  cohen_threshold <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  colorPalette <span class="op">=</span> <span class="st">"Set3"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame of normalized (non-transformed) counts where each row is a gene and each column is a sample.
Row names should contain gene names, and column names should contain sample identifiers. <strong>(Required)</strong></p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>A data frame describing the attributes of each sample, where each row corresponds to a sample and each column to an attribute.
The first column should contain sample identifiers (i.e., the column names of <code>data</code>). <strong>(Required if method = "all")</strong></p></dd>


<dt id="arg-gene-sets">gene_sets<a class="anchor" aria-label="anchor" href="#arg-gene-sets"></a></dt>
<dd><p>Gene set input. <strong>(Required)</strong></p><ul><li><p><strong>Unidirectional gene sets</strong>: Provide a named list where each element is a vector of gene names representing a gene signature.</p></li>
<li><p><strong>Bidirectional gene sets</strong>: Provide a named list where each element is a data frame with two columns:</p><ul><li><p>The <strong>first column</strong> contains gene names.</p></li>
<li><p>The <strong>second column</strong> indicates the expected direction of enrichment (1 for upregulated genes, -1 for downregulated genes).</p></li>
</ul></li>
</ul></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string indicating the scoring method to use. Options are <code>"ssGSEA"</code>, <code>"logmedian"</code> or <code>"ranking"</code>. Defaults to <code>"logmedian"</code>.</p></dd>


<dt id="arg-colorvariable">ColorVariable<a class="anchor" aria-label="anchor" href="#arg-colorvariable"></a></dt>
<dd><p>Optional. Name of the metadata column to use for point color in plots.</p></dd>


<dt id="arg-groupingvariable">GroupingVariable<a class="anchor" aria-label="anchor" href="#arg-groupingvariable"></a></dt>
<dd><p>Optional. Name of the metadata column to use for group comparison.</p></dd>


<dt id="arg-colorvalues">ColorValues<a class="anchor" aria-label="anchor" href="#arg-colorvalues"></a></dt>
<dd><p>Optional. Named vector of colors to use for each group in <code>ColorVariable</code> or <code>GroupingVariable</code>.</p></dd>


<dt id="arg-connectgroups">ConnectGroups<a class="anchor" aria-label="anchor" href="#arg-connectgroups"></a></dt>
<dd><p>Logical. If TRUE, connects points of the same sample across conditions.</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>Number of columns in the facet layout of the plot.</p></dd>


<dt id="arg-nrow">nrow<a class="anchor" aria-label="anchor" href="#arg-nrow"></a></dt>
<dd><p>Number of rows in the facet layout of the plot.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>Optional. Main title of the plot.</p></dd>


<dt id="arg-widthtitle">widthTitle<a class="anchor" aria-label="anchor" href="#arg-widthtitle"></a></dt>
<dd><p>Numeric. Width of the title area (for alignment purposes).</p></dd>


<dt id="arg-titlesize">titlesize<a class="anchor" aria-label="anchor" href="#arg-titlesize"></a></dt>
<dd><p>Numeric. Font size of the title text.</p></dd>


<dt id="arg-limits">limits<a class="anchor" aria-label="anchor" href="#arg-limits"></a></dt>
<dd><p>Optional numeric vector of length 2 specifying y-axis limits.</p></dd>


<dt id="arg-legend-nrow">legend_nrow<a class="anchor" aria-label="anchor" href="#arg-legend-nrow"></a></dt>
<dd><p>Optional. Number of rows in the plot legend.</p></dd>


<dt id="arg-pointsize">pointSize<a class="anchor" aria-label="anchor" href="#arg-pointsize"></a></dt>
<dd><p>Numeric. Size of the points in the plots.</p></dd>


<dt id="arg-xlab">xlab<a class="anchor" aria-label="anchor" href="#arg-xlab"></a></dt>
<dd><p>Optional. Label for the x-axis.</p></dd>


<dt id="arg-labsize">labsize<a class="anchor" aria-label="anchor" href="#arg-labsize"></a></dt>
<dd><p>Numeric. Font size for axis and facet labels.</p></dd>


<dt id="arg-compute-cohen">compute_cohen<a class="anchor" aria-label="anchor" href="#arg-compute-cohen"></a></dt>
<dd><p>Logical. If TRUE, computes Cohen's d effect sizes between groups.</p></dd>


<dt id="arg-cond-cohend">cond_cohend<a class="anchor" aria-label="anchor" href="#arg-cond-cohend"></a></dt>
<dd><p>Optional. Specify a condition or comparison subset for calculating Cohen’s d.</p></dd>


<dt id="arg-pvalcalc">pvalcalc<a class="anchor" aria-label="anchor" href="#arg-pvalcalc"></a></dt>
<dd><p>Logical. If TRUE, computes p-values for group comparisons.</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>Character string indicating comparison complexity. Options: <code>"simple"</code>, <code>"medium"</code>, <code>"extensive"</code>.</p></dd>


<dt id="arg-widthlegend">widthlegend<a class="anchor" aria-label="anchor" href="#arg-widthlegend"></a></dt>
<dd><p>Numeric. Width of the legend area in volcano plots.</p></dd>


<dt id="arg-cohen-threshold">cohen_threshold<a class="anchor" aria-label="anchor" href="#arg-cohen-threshold"></a></dt>
<dd><p>Numeric. Cohen's d threshold to highlight effect size in volcano plots (default = 0.6).</p></dd>


<dt id="arg-colorpalette">colorPalette<a class="anchor" aria-label="anchor" href="#arg-colorpalette"></a></dt>
<dd><p>Character. Name of RColorBrewer palette for coloring (default = <code>"Set3"</code>).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Four methods are available:</p><ul><li><p><strong>ssGSEA</strong>: Uses the single-sample Gene Set Enrichment Analysis (ssGSEA) method to compute an enrichment score
for each signature in each sample using an adaptation of the <code>gsva()</code> function from the <code>GSVA</code> package.</p></li>
<li><p><strong>logmedian</strong>: Computes the score as the sum of the normalized (log2-median-centered) expression values of the
signature genes divided by the number of genes in the signature.</p></li>
<li><p><strong>ranking</strong>: Computes gene signature scores for each sample by ranking the expression of signature genes
in the dataset and normalizing the score based on the total number of genes.</p></li>
<li><p><strong>all</strong>: Computes gene signature scores using all three methods (<code>ssGSEA</code>, <code>logmedian</code>, and <code>ranking</code>).
Returns a heatmap summarizing Cohen's D for all metric combinations of the variables of interest.</p></li>
</ul><p>Depending on the method and the type of variable (categorical, numeric, or <code>NULL</code>), the function produces different plots:</p><ul><li><p><strong>If <code>method = "all"</code></strong> and the variable is <strong>categorical</strong>, a heatmap of Cohen's d or F statistics and a volcano plot showing contrasts between all groups of that variable are produced.</p></li>
<li><p><strong>If <code>method = "all"</code></strong> and the variable is <strong>numeric</strong>, a heatmap of Cohen's F and a volcano plot are produced.</p></li>
<li><p><strong>If <code>method != "all"</code></strong> and the variable is <strong>categorical</strong>, a violin plot for each signature is generated.</p></li>
<li><p><strong>If <code>method != "all"</code></strong> and the variable is <code>NULL</code>, a density plot of the score distribution is displayed.</p></li>
<li><p><strong>If <code>method != "all"</code></strong> and the variable is <strong>numeric</strong>, a scatter plot is created to show the relationship between the scores and the numeric variable.</p></li>
<li><p><strong>If <code>method = "all"</code></strong> and the variable is <strong>categorical</strong>, the function returns a heatmap of Cohen's d or F statistics and a volcano plot showing contrasts between all groups of that variable.</p></li>
<li><p><strong>If <code>method = "all"</code></strong> and the variable is <strong>numeric</strong>, a heatmap of Cohen's F and a volcano plot will be produced.</p></li>
<li><p><strong>If <code>method != "all"</code></strong> and the variable is <strong>categorical</strong>, a violin plot for each signature will be displayed.</p></li>
<li><p><strong>If <code>method != "all"</code></strong> and the variable is <code>NULL</code>, a density plot of the score distribution will be displayed.</p></li>
<li><p><strong>If <code>method != "all"</code></strong> and the variable is <strong>numeric</strong>, a scatter plot will be generated to show the relationship between the scores and the numeric variable.</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rita Martins-Silva, Alexandre Kaizeler, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

