<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Score Variable Association — Score_VariableAssociation • markeR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Score Variable Association — Score_VariableAssociation"><meta name="description" content="This function evaluates the association between gene expression scores and metadata variables.
It uses linear modeling to get Cohen's F, and contrast-based comparisons for categorical variables to compute Cohen's D.
The function generates plots summarizing the results."><meta property="og:description" content="This function evaluates the association between gene expression scores and metadata variables.
It uses linear modeling to get Cohen's F, and contrast-based comparisons for categorical variables to compute Cohen's D.
The function generates plots summarizing the results."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">markeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseTranscriptomicsLab/markeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Score Variable Association</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseTranscriptomicsLab/markeR/blob/main/R/Score_VariableAssociation.R" class="external-link"><code>R/Score_VariableAssociation.R</code></a></small>
      <div class="d-none name"><code>Score_VariableAssociation.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function evaluates the association between gene expression scores and metadata variables.
It uses linear modeling to get Cohen's F, and contrast-based comparisons for categorical variables to compute Cohen's D.
The function generates plots summarizing the results.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Score_VariableAssociation</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  <span class="va">cols</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logmedian"</span>, <span class="st">"ssGSEA"</span>, <span class="st">"ranking"</span><span class="op">)</span>,</span>
<span>  <span class="va">gene_set</span>,</span>
<span>  mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"simple"</span>, <span class="st">"medium"</span>, <span class="st">"extensive"</span><span class="op">)</span>,</span>
<span>  nonsignif_color <span class="op">=</span> <span class="st">"grey"</span>,</span>
<span>  signif_color <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>  saturation_value <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sig_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  widthlabels <span class="op">=</span> <span class="fl">18</span>,</span>
<span>  labsize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  titlesize <span class="op">=</span> <span class="fl">14</span>,</span>
<span>  pointSize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  discrete_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  continuous_color <span class="op">=</span> <span class="st">"#8C6D03"</span>,</span>
<span>  color_palette <span class="op">=</span> <span class="st">"Set2"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame or matrix containing gene expression data.</p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>A data frame containing sample metadata with at least one column corresponding to the variables of interest.</p></dd>


<dt id="arg-cols">cols<a class="anchor" aria-label="anchor" href="#arg-cols"></a></dt>
<dd><p>A character vector specifying metadata columns to analyse.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string specifying the scoring method (<code>"logmedian"</code>, <code>"ssGSEA"</code>, or <code>"ranking"</code>).</p></dd>


<dt id="arg-gene-set">gene_set<a class="anchor" aria-label="anchor" href="#arg-gene-set"></a></dt>
<dd><p>A named list containing one gene set for scoring.</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>A character string specifying the contrast generation method (<code>"simple"</code>, <code>"medium"</code>, <code>"extensive"</code>). Four methods are available:</p><ul><li><p><strong>ssGSEA</strong>: Uses the single-sample Gene Set Enrichment Analysis (ssGSEA) method to compute an enrichment score
for each signature in each sample using an adaptation of the <code>gsva()</code> function from the <code>GSVA</code> package.</p></li>
<li><p><strong>logmedian</strong>: Computes the score as the sum of the normalized (log2-median-centered) expression values of the
signature genes divided by the number of genes in the signature.</p></li>
<li><p><strong>ranking</strong>: Computes gene signature scores for each sample by ranking the expression of signature genes
in the dataset and normalizing the score based on the total number of genes.</p></li>
</ul></dd>


<dt id="arg-nonsignif-color">nonsignif_color<a class="anchor" aria-label="anchor" href="#arg-nonsignif-color"></a></dt>
<dd><p>A string specifying the color for non-significant results. Default: <code>"grey"</code>.</p></dd>


<dt id="arg-signif-color">signif_color<a class="anchor" aria-label="anchor" href="#arg-signif-color"></a></dt>
<dd><p>A string specifying the color for significant results. Default: <code>"red"</code>.</p></dd>


<dt id="arg-saturation-value">saturation_value<a class="anchor" aria-label="anchor" href="#arg-saturation-value"></a></dt>
<dd><p>A numeric value for color saturation threshold. Default: <code>NULL</code> (auto-determined).</p></dd>


<dt id="arg-sig-threshold">sig_threshold<a class="anchor" aria-label="anchor" href="#arg-sig-threshold"></a></dt>
<dd><p>A numeric value specifying the significance threshold. Default: <code>0.05</code>.</p></dd>


<dt id="arg-widthlabels">widthlabels<a class="anchor" aria-label="anchor" href="#arg-widthlabels"></a></dt>
<dd><p>An integer controlling contrast label wrapping. Default: <code>18</code>.</p></dd>


<dt id="arg-labsize">labsize<a class="anchor" aria-label="anchor" href="#arg-labsize"></a></dt>
<dd><p>An integer controlling axis text size. Default: <code>10</code>.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>A string specifying the plot title. Default: <code>NULL</code>.</p></dd>


<dt id="arg-titlesize">titlesize<a class="anchor" aria-label="anchor" href="#arg-titlesize"></a></dt>
<dd><p>An integer specifying the title size. Default: <code>14</code>.</p></dd>


<dt id="arg-pointsize">pointSize<a class="anchor" aria-label="anchor" href="#arg-pointsize"></a></dt>
<dd><p>A numeric value for point size in plots. Default: <code>5</code>.</p></dd>


<dt id="arg-discrete-colors">discrete_colors<a class="anchor" aria-label="anchor" href="#arg-discrete-colors"></a></dt>
<dd><p>A named list mapping categorical variable levels to colors.
Each element should be a named vector where names correspond to factor levels. Default: <code>NULL</code>.</p></dd>


<dt id="arg-continuous-color">continuous_color<a class="anchor" aria-label="anchor" href="#arg-continuous-color"></a></dt>
<dd><p>A string specifying the color for continuous variables. Default: <code>"#8C6D03"</code>.</p></dd>


<dt id="arg-color-palette">color_palette<a class="anchor" aria-label="anchor" href="#arg-color-palette"></a></dt>
<dd><p>A string specifying the color palette for discrete variables. Default: <code>"Set2"</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with:</p><ul><li><p><code>Overall</code>: Data frame of effect sizes and p-values for each contrasted phenotypic variable.</p></li>
<li><p><code>Contrasts</code>: Data frame of Cohen’s d and adjusted p-values for contrasts between levels of categorical variables,
with the resolution of contrasts determined by the mode parameter.</p></li>
<li><p><code>plot</code>: A combined visualization with three main panels:
(1) lollipop plots of Cohen’s f for each variable of interest,
(2) distribution plots of the score by variable (density or scatter depending on variable type),
and (3, if applicable) lollipop plots of Cohen’s d for contrasts in categorical variables.</p></li>
<li><p><code>plot_contrasts</code>: Lollipop plots of Cohen’s d effect sizes for contrasts between levels of non numerical variables (if applicable), colored by adjusted p-value (BH).</p></li>
<li><p><code>plot_overall</code>: Lollipop plot showing Cohen’s f effect sizes for each variable, colored by p-value.</p></li>
<li><p><code>plot_distributions</code>: List of density or scatter plots of the score across variable levels, depending on variable type.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span>  <span class="co"># Name columns as Gene1, Gene2, ..., Gene10</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>  <span class="co"># Name rows as Sample1, Sample2, ..., Sample100</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>,  <span class="co"># Sample ID matches the rownames of the data</span></span></span>
<span class="r-in"><span>  Condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>  <span class="co"># Two conditions (A and B)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SampleSet <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">Score_VariableAssociation</span><span class="op">(</span><span class="va">data</span>, <span class="va">metadata</span>, cols <span class="op">=</span> <span class="st">"Condition"</span>, gene_set <span class="op">=</span> <span class="va">gene_set</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Considering unidirectional gene signature mode for signature SampleSet</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>no non-missing arguments to min; returning Inf</span>
<span class="r-plt img"><img src="Score_VariableAssociation-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">plot</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="Score_VariableAssociation-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rita Martins-Silva, Alexandre Kaizeler, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

