<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Visualize Individual Genes' Expression Using Various Methods — VisualiseIndividualGenes • markeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Visualize Individual Genes' Expression Using Various Methods — VisualiseIndividualGenes"><meta name="description" content="This function provides multiple visualization methods for analyzing individual genes
from an expression dataset. Users can generate violin plots, expression heatmaps,
correlation heatmaps, ROC curves, and effect size plots."><meta property="og:description" content="This function provides multiple visualization methods for analyzing individual genes
from an expression dataset. Users can generate violin plots, expression heatmaps,
correlation heatmaps, ROC curves, and effect size plots."><meta property="og:image" content="https://diseasetranscriptomicslab.github.io/markeR/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">markeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseTranscriptomicsLab/markeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Visualize Individual Genes' Expression Using Various Methods</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseTranscriptomicsLab/markeR/blob/main/R/VisualiseIndividualGenes.R" class="external-link"><code>R/VisualiseIndividualGenes.R</code></a></small>
      <div class="d-none name"><code>VisualiseIndividualGenes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function provides multiple visualization methods for analyzing individual genes
from an expression dataset. Users can generate violin plots, expression heatmaps,
correlation heatmaps, ROC curves, and effect size plots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">VisualiseIndividualGenes</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  <span class="va">genes</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"violins"</span>, <span class="st">"expr_heatmap"</span>, <span class="st">"corr_heatmap"</span>, <span class="st">"roc"</span>, <span class="st">"effectsize"</span>, <span class="st">"all"</span><span class="op">)</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  GroupingVariable <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ColorVariable <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ColorValues <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pointSize <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  widthTitle <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  y_limits <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  divide <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  invert_divide <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  xlab <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colorlab <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame where rows correspond to genes and columns correspond to samples.
The row names should be gene identifiers.</p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>A data frame containing sample metadata. The first column should contain
sample identifiers matching the column names of <code>data</code>.</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>A character vector of gene names to be plotted. If more than 20 genes are provided, a warning is issued.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string specifying which visualization method to use.
Options are <code>"violins"</code>, <code>"expr_heatmap"</code>, <code>"corr_heatmap"</code>, <code>"roc"</code>, <code>"effectsize"</code>, or <code>"all"</code>.
Default is <code>"violins"</code>. If <code>"all"</code> is selected, all methods will be applied.</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>A logical value indicating whether to display the generated plots (<code>TRUE</code>) or not (<code>FALSE</code>).
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-groupingvariable">GroupingVariable<a class="anchor" aria-label="anchor" href="#arg-groupingvariable"></a></dt>
<dd><p>A character string specifying the column name in <code>metadata</code> to be used
for grouping samples on the x-axis in violin plots.</p></dd>


<dt id="arg-colorvariable">ColorVariable<a class="anchor" aria-label="anchor" href="#arg-colorvariable"></a></dt>
<dd><p>A character string specifying the column name in <code>metadata</code> used
to color the points in violin plots.</p></dd>


<dt id="arg-colorvalues">ColorValues<a class="anchor" aria-label="anchor" href="#arg-colorvalues"></a></dt>
<dd><p>An optional named vector mapping unique values of <code>ColorVariable</code> to specific colors.</p></dd>


<dt id="arg-pointsize">pointSize<a class="anchor" aria-label="anchor" href="#arg-pointsize"></a></dt>
<dd><p>A numeric value specifying the size of jittered points in violin plots. Default is <code>2</code>.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>An optional character string specifying the title of the plot.</p></dd>


<dt id="arg-widthtitle">widthTitle<a class="anchor" aria-label="anchor" href="#arg-widthtitle"></a></dt>
<dd><p>An optional integer specifying the maximum width of the title before inserting line breaks. Default is <code>16</code>.</p></dd>


<dt id="arg-y-limits">y_limits<a class="anchor" aria-label="anchor" href="#arg-y-limits"></a></dt>
<dd><p>A numeric vector of length 2 specifying the limits of the y-axis in violin plots.
If <code>NULL</code>, the y-axis is adjusted automatically. Default is <code>NULL</code>.</p></dd>


<dt id="arg-legend-nrow">legend_nrow<a class="anchor" aria-label="anchor" href="#arg-legend-nrow"></a></dt>
<dd><p>A numeric value specifying the number of rows in the legend for violin plots.
If <code>NULL</code>, the default number of rows is determined automatically.</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>An optional numeric value specifying the number of columns in the grid layout for violin plots.
If <code>NULL</code>, it is automatically determined.</p></dd>


<dt id="arg-nrow">nrow<a class="anchor" aria-label="anchor" href="#arg-nrow"></a></dt>
<dd><p>An optional numeric value specifying the number of rows in the grid layout for violin plots.
If <code>NULL</code>, it is automatically determined.</p></dd>


<dt id="arg-divide">divide<a class="anchor" aria-label="anchor" href="#arg-divide"></a></dt>
<dd><p>A character string specifying a column in <code>metadata</code> used to divide the violin plots into facets.
Default is <code>NULL</code>.</p></dd>


<dt id="arg-invert-divide">invert_divide<a class="anchor" aria-label="anchor" href="#arg-invert-divide"></a></dt>
<dd><p>A logical value indicating whether to invert the facet layout of violin plots.
Default is <code>FALSE</code>.</p></dd>


<dt id="arg-xlab">xlab<a class="anchor" aria-label="anchor" href="#arg-xlab"></a></dt>
<dd><p>An optional character string specifying the label for the x-axis in violin plots.
Default is the name of <code>GroupingVariable</code>.</p></dd>


<dt id="arg-colorlab">colorlab<a class="anchor" aria-label="anchor" href="#arg-colorlab"></a></dt>
<dd><p>An optional character string specifying the legend title for the color variable.
Default is <code>NULL</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A named list where each selected method (e.g., <code>"violins"</code>, <code>"expr_heatmap"</code>, etc.) is a key.
Each element in this list is itself a named list with the following structure:</p><dl><dt><code>data</code></dt>
<dd><p>A data frame or matrix containing the transformed data used for plotting, formatted according to the selected method.</p></dd>

<dt><code>plot</code></dt>
<dd><p>The generated plot object (e.g., a ggplot2 object) corresponding to the visualization.</p></dd>

<dt><code>aux</code></dt>
<dd><p>Additional parameters, specific of each method.</p></dd>


</dl><p>If <code>plot = TRUE</code>, the plots are also displayed in the output.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function calls multiple sub-functions based on the selected visualization method:</p><ul><li><p><code><a href="IndividualGenes_Violins.html">IndividualGenes_Violins()</a></code>: Generates violin plots for gene expression (log2 transformed).</p></li>
<li><p><code>IndividualGenes_ExprHeatmap()</code>: Creates a heatmap of expression values.</p></li>
<li><p><code>IndividualGenes_CorrHeatmap()</code>: Computes and visualizes a correlation heatmap.</p></li>
<li><p><code>IndividualGenes_ROC()</code>: Plots receiver operating characteristic (ROC) curves.</p></li>
<li><p><code>IndividualGenes_EffectSize()</code>: Computes and visualizes effect sizes.</p></li>
</ul><p>The function returns a list of plots. If <code>method = "all"</code>, all available visualizations are included.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rita Martins-Silva, Alexandre Kaizeler, Nuno Luís Barbosa-Morais.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

